p4-16/psa/PSA.mdk

```diff
@@ -450,11 +450,52 @@ Parameters:
 
 ## Counters
 
-Counters are a simple mechanism for keeping statistics about the
-packets that trigger a table in a Match Action unit.
-
-Direct counters fire when the count method is invoked in an action, and
-have an instance for each entry in the table.
+Counters are a mechanism for keeping statistics.  The control plane
+can read counter values.  A P4 program cannot read counter values,
+only update them.  If you wish to implement a feature involving
+sequence numbers in packets, for example, use Registers instead
+(Section [#sec-registers]).
+
+Direct counters are counters associated with a particular P4 table,
+and are implemented by the extern `DirectCounter`.  There are also
+indirect counters, which are implemented by the extern `Counter` (the
+name 'indirect' does not imply that there is any indirection involved
+in the implementation, only that they are not direct counters).  The
+primary differences between direct counters and indirect counters are:
+
+- Number of independently updatable counter values:
+  - A single instantiation of a direct counter always contains as many
+    independent counter values as the number of entries in the table
+    with which it is associated (TBD: see below for what this means
+    for tables that use action profiles).
+  - You must specify the number of independent counter values for an
+    indirect counter when instantiating it.  This number of counters
+    need not be the same as the size of any table.
+- Where counter updates are allowed in the P4 program:
+  - For a direct counter, you may only invoke its `count` method from
+    inside the actions of the table with which it is associated, and
+    this always updates the counter value associated with the matching
+    table entry.
+  - For an indirect counter, you may invoke its `count` method
+    anywhere in the P4 program where extern object method invocations
+    are permitted (e.g. inside actions, or directly inside a control's
+    `apply` block), and every such invocation must specify the index
+    of the counter value to be updated.
+
+Counters are only intended to support packet counters and byte
```

Comment:
Are we mandating the bitwidth of a counter? If not, should it be a parameter of the extern constructor?